### ISSUE TITLE
[Hold for #23757][$1000] Web - LHN - Emoji picker doesn't close when navigating to other chat with the browser back button

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#24559]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat with Concierge or other chat
2. Then proceed to open a chat with user B
3. Click on emoji icon from composer
4. Click on the browser's back button
5. Click on one of the emojis
6. Open the chat you started before starting chat with user B

## Expected Result:
Emoji picker should either close on navigating to other chat or the selected emoji should be entered to the current composer

## Actual Result:
Emoji picker doesn't close & selecting emoji inserts emoji user B's composer

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.54-2

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856 

**Notes/Photos/Videos:** Any additional supporting documentation 

https://github.com/Expensify/App/assets/115492554/b71a697e-444f-4ba9-bfe8-488c133bac42

https://github.com/Expensify/App/assets/115492554/0af2460a-a457-4379-9d22-8c1bc16e4df7

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Natnael-Guchima

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691182964868429

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a49af3096640d7b0</li>
        <li>Upwork Job ID: 1693415547996028928</li>
        <li>Last Price Increase: 2023-08-21</li>
    </ul>
</details>

**Comment by Pluto0104** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Emoji picker doesn't close when navigating to other chat with the browser back button.

### What is the root cause of that problem?
The app does not handle emoji picker when navigating to other chat with the browser back button.

### What changes do you think we should make in order to solve the problem?
https://github.com/Expensify/App/blob/main/src/components/EmojiPicker/EmojiPicker.js
As we can see this component, we don't have any functions for navigating.
So we need to define the listener for navigating.
For example:
```
const onChangeRoute = () => {
    if (!EmojiPickerAction.emojiPickerRef.current || !EmojiPickerAction.isEmojiPickerVisible) {
        return;
    }
    EmojiPickerAction.hideEmojiPicker();
};
```

And then we need to pass this prop to `NavigationRoot` component.
https://github.com/Expensify/App/blob/6d1b9922effa021056e208407da1f89f163b8771/src/Expensify.js#L210C23-L212

For example:
```
<NavigationRoot
    onChangeRoute={onChangeRoute}
    onReady={setNavigationReady}
    authenticated={isAuthenticated}
 />
```

And then in `NavigationRoot` component:
```
const updateSavedNavigationStateAndLogRoute = (state) => {
    if (!state) {
        return;
    }
    props.onChangeRoute();
    navigationStateRef.current = state;
    updateCurrentReportID(state);
    parseAndLogRoute(state);
    animateStatusBarBackgroundColor();
};
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by jfquevedol2198** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Web - LHN - Emoji picker doesn't close when navigating to other chat with the browser back button

### What is the root cause of that problem?
When `currentReportID` is changed, we are not closing EmojiPicker.

### What changes do you think we should make in order to solve the problem?
Add code to hide `EmojiPicker` when `currentReportID` is changed in [EmojiPickerButton.js](https://github.com/Expensify/App/blob/4230f289341c8d6d9a8b780d54da0725d4764b2b/src/components/EmojiPicker/EmojiPickerButton.js)
```javascript
    const {currentReportID} = useCurrentReportID();
    useEffect(() => {
        if (!EmojiPickerAction.emojiPickerRef.current || !EmojiPickerAction.emojiPickerRef.current.isEmojiPickerVisible) return;
        EmojiPickerAction.emojiPickerRef.current.hideEmojiPicker();
    }, [currentReportID])
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by kadiealexander** [comment]
@Natnael-Guchima I'm having a lot of trouble following your reproduction steps. It looks like, from the video, you're logged into two accounts. Can you please write more detailed steps that make it clear which user and tab to use?

**Comment by ShogunFire** [comment]
@kadiealexander No need for two accounts to reproduce.

**Comment by Natnael-Guchima** [comment]
@kadiealexander please try the steps listed below, and let me know if you are not able to reproduce the issue. 

**Action Performed:** 
1. Open a chat with Concierge or another chat
2. Then proceed to open a chat with user B
3. Click on emoji icon from composer
4. Click on the browser's back button

**Expected Result:** 
Emoji picker should close when navigating to other chat 


https://github.com/Expensify/App/assets/57330233/964ee030-4307-414a-93d9-4adb4d9e8a72

**Comment by kadiealexander** [comment]
I can't reproduce on the latest version:

https://github.com/Expensify/App/assets/59587260/357fcd8f-3602-4633-8383-ea20f2753552

@Natnael-Guchima could you please test on version v1.3.54-11 and let me know if you're still experiencing it?

**Comment by Natnael-Guchima** [comment]
I am still able to reproduce it on the latest version (v1.3.54-11). I am not sure if this is a device-specific issue.  @ShogunFire are you able to reproduce this on that latest release? 

https://github.com/Expensify/App/assets/57330233/dd0a4c80-c31c-4554-9d2f-196b1f4c54f8

**Comment by kadiealexander** [comment]
@Natnael-Guchima are you testing on Windows? It's possible it's windows-specific.

**Comment by Natnael-Guchima** [comment]
Yes, I am testing on windows.

**Comment by ShogunFire** [comment]
Also testing on windows

**Comment by jfquevedol2198** [comment]
I can reproduce this on staging (v1.3.54-11) and production (v1.3.53-2)
cc @ShogunFire @kadiealexander @Natnael-Guchima

**Comment by jfquevedol2198** [comment]
MacOS / Chrome
cc @Natnael-Guchima

**Comment by jfquevedol2198** [comment]
@Natnael-Guchima I can also reproduce this on Windows

**Comment by kadiealexander** [comment]
I was able to repro this morning on Mac Chrome v1.3.54-13.

**Comment by Natnael-Guchima** [comment]
@kadiealexander this issue is fixed on dev now. The fixing PR is [here](https://github.com/Expensify/App/pull/24479) which was intended to fix another issue that is reported [here](https://github.com/Expensify/App/issues/23757). The discussion to fix this issue - the one I reported - on the fixing PR starts [here](https://github.com/Expensify/App/pull/24479#issuecomment-1678462899) on Aug 15, and I reported this issue on Aug 5. On the fixing PR, a new solution was introduced [here](https://github.com/Expensify/App/pull/24479#issuecomment-1678500203) to fix the issue I reported on Aug 15:  
>The new solution will also fix https://github.com/Expensify/App/issues/24559 (btw, do we need to compensate @hoangzinh too?).

_This is to ask if I am eligible for reporting bonus since the discussion to fix the issue I reported comes after the issue I reported? 
Thanks._ 


 
The issue fixed on dev: 

https://github.com/Expensify/App/assets/57330233/fff641e2-ff48-47e7-96e3-b951219f7a06

The current behavior of the issue on the latest staging. 

https://github.com/Expensify/App/assets/57330233/f34ca287-a82b-4bc7-944a-e77e726b2781

My reporting date 

![Screenshot from 2023-08-25 09-32-10](https://github.com/Expensify/App/assets/57330233/7c872f6b-c703-44fe-a32c-94e3746ffacc)

**Comment by kadiealexander** [comment]
@Natnael-Guchima I don't think we typically pay out reporting bonuses for issues that were fixed elsewhere. Generally if something goes on hold because another PR will fix it, we don't make any payments when it comes off hold and is fixed. Sorry about that!
